modules/internal/String.chpl

```diff
@@ -247,44 +294,181 @@ module String {
     return codepointIndex;
   }
 
-  // Helper routines in support of being able to use ranges of codepointIndex
+  // Helper routines in support of being able to use ranges of indices
   pragma "no doc"
-  proc chpl_build_bounded_range(low: codepointIndex, high: codepointIndex)
-    return new range(codepointIndex, _low=low, _high=high);
+  proc chpl_build_bounded_range(low: ?t, high: t)
+    where t == byteIndex || t == codepointIndex
+    return new range(t, _low=low, _high=high);
```

Comment:
This is fine, but AFAIK there is a compile-time advantage to constraining types in the argument type (like `low:codepointIndex`) instead of using a where clause. For that reason, I'd probably just duplicate this code, once for `codepointIndex` and once for `byteIndex`.

It's OK with me if you decide not to do this.

Reply:
Thanks.  For now, I've decided to keep these functions generic because there are 27 of them per type.  That expands out to 54 now, and 81 later if we add graphemes, so it significantly affects readability by humans.

I'm open to revisiting this decision later, though, since I'm sensitive to the compile speed issue.

Reply:
Now that initializers are coming up to snuff, it seems like these routines should get replaced with initializers which may be able to be more generic or wrestle with the type questions internally while also not requiring a bunch of distinctly named routines (which I think is similar to what bharsh has been doing with domains and such lately).  Not that I'm suggesting doing that in this PR.